### PR TITLE
Add specification for type synonyms in DAML-LF

### PR DESCRIPTION
**Pull request by cocreature**
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags, if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/spec/daml-lf-1.rst [hunk 1]
```diff
@@ -676,6 +681,7 @@ available for usage::
        |  'variant' T (α₁: k₁)… (αₙ: kₙ) ↦ V₁ : τ₁ | … | Vₘ : τₘ
                                                     -- DefVariant
        |  'enum' T  ↦ E₁ | … | Eₘ                    -- DefEnum
+       |  'synonym' T (α₁: k₁)… (αₙ: kₙ) ↦ τ        -- DefTypeSynonym
```
Comment (hurryabit) [inline daml-lf/spec/daml-lf-1.rst]:
```suggestion
       |  'synonym' S (α₁: k₁)… (αₙ: kₙ) ↦ τ        -- DefTypeSynonym
```

File: daml-lf/spec/daml-lf-1.rst [hunk 2]
```diff
@@ -846,6 +852,14 @@ First, we formally defined *well-formed types*. ::
     ————————————————————————————————————————————— TyTuple
       Γ  ⊢  ⟨ f₁: τ₁, …, fₙ: τₙ ⟩  :  ⋆
 
+      'synonym' T (α₁:k₁) … (αₙ:kₙ) ↦ τ ∈ 〚Ξ〛Mod
```
Comment (hurryabit) [inline daml-lf/spec/daml-lf-1.rst]:
```suggestion
      'synonym' S (α₁:k₁) … (αₙ:kₙ) ↦ τ ∈ 〚Ξ〛Mod
```

**Comment by remyhaemmerle-da** [comment]
This PR advances the state of #3616.

File: daml-lf/spec/daml-lf-1.rst [hunk 3]
```diff
@@ -846,6 +852,14 @@ First, we formally defined *well-formed types*. ::
     ————————————————————————————————————————————— TyTuple
       Γ  ⊢  ⟨ f₁: τ₁, …, fₙ: τₙ ⟩  :  ⋆
 
+      'synonym' S (α₁:k₁) … (αₙ:kₙ) ↦ τ ∈ 〚Ξ〛Mod
+      (α₁:k₁) … (αₙ:kₙ) · Γ  ⊢  τ  :  ⋆
```
Comment (remyhaemmerle-da) [inline daml-lf/spec/daml-lf-1.rst]:
```suggestion
```

This line would better go in Well-formed definitions paragraph.

```
    αₙ : kₙ · ⋯ · α₁ : k₁  ⊢  τ  :  ⋆
  ——————————————————————————————————————————————————————————————— DefSyn
    ⊢  'synonym' S (α₁: k₁) … (αₙ: kₙ) ↦ τ 
```

File: daml-lf/spec/daml-lf-1.rst [hunk 4]
```diff
@@ -846,6 +852,14 @@ First, we formally defined *well-formed types*. ::
...
+      Γ  ⊢  τ₁ : k₁  …  Γ  ⊢  τₙ : kₙ
+    ————————————————————————————————————————————— TyTypeSynonym
+      Γ  ⊢  S τ₁ … τₙ  :  ⋆
```
Comment (remyhaemmerle-da) [inline daml-lf/spec/daml-lf-1.rst]:
```suggestion
      Γ  ⊢  Mod:S τ₁ … τₙ  :  ⋆
```

File: daml-lf/spec/daml-lf-1.rst [hunk 5]
```diff
@@ -982,6 +996,12 @@ Then we define *well-formed expressions*. ::
     ——————————————————————————————————————————————————————————————— ExpEnumCon
       Γ  ⊢  Mod:T:Eᵢ  :  Mod:T
 
+      'synonym' S (α₁:k₁) … (αₙ:kₙ) ↦ τ
+      Γ  ⊢  τ₁ : k₁    ⋯     Γ  ⊢  τₙ : kₙ
```
Comment (remyhaemmerle-da) [inline daml-lf/spec/daml-lf-1.rst]:
```suggestion
      Γ  ⊢  τ₁ : k₁    ⋯     Γ  ⊢  τₙ : kₙ
```
Rules for expression assumes _well-formed types_

File: daml-lf/spec/daml-lf-1.rst (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/spec/daml-lf-1.rst]: Apparently it was not fixed.

**Pull request merged**